server/test/Test/Utopia/Web/Database.hs

```diff
@@ -0,0 +1,69 @@
+import           Data.Time.Clock
+import           GHC.Conc
+import           Prelude                        (String)
+import           Protolude
+import           System.Timeout
+import           Test.HUnit.Lang
+import           Test.Tasty
+import           Test.Tasty.HUnit
+import           Test.Utopia.Web.Database.Utils
+import           Test.Utopia.Web.Executors.Test
+import           Utopia.ClientModel
+import           Utopia.Web.Database
+import           Utopia.Web.Database.Types
+import           Utopia.Web.Executors.Common
+
+sixteenthOfJanuaryMorning :: IO UTCTime
+sixteenthOfJanuaryMorning = pure $ UTCTime (fromOrdinalDate 2024 16) (secondsToDiffTime (9 * 60 * 60))
+
+sixteenthOfJanuaryAfternoon :: IO UTCTime
+sixteenthOfJanuaryAfternoon = pure $ UTCTime (fromOrdinalDate 2024 16) (secondsToDiffTime (13 * 60 * 60))
+
+controlSpec :: Bool -> TestTree
+controlSpec enableExternalTests =
+  let toggledGroup groupName tests = if enableExternalTests then sequentialTestGroup groupName AllSucceed tests else testGroup groupName []
+  in  toggledGroup "Project control"
+        [ withTestPool "can release control when control is not held" $ \pool -> do
+            metrics <- getDatabaseMetrics
+            releaseCollaborationControl metrics pool "testowner" "testproject" "testeditor"
+        , withTestPool "can release control when control is held" $ \pool -> do
+            metrics <- getDatabaseMetrics
+            forceClaimCollaborationControl metrics pool sixteenthOfJanuaryMorning "testowner" "testproject" "testeditor"
+            releaseCollaborationControl metrics pool "testowner" "testproject" "testeditor"
+        , withTestPool "when one user claims control, another cannot" $ \pool -> do
+            metrics <- getDatabaseMetrics
+            firstClaimResult <- maybeClaimCollaborationControl metrics pool sixteenthOfJanuaryMorning "firstowner" "testproject" "firsteditor"
+            assertBool "First claim result should be successful." firstClaimResult
+            secondClaimResult <- maybeClaimCollaborationControl metrics pool sixteenthOfJanuaryMorning "secondowner" "testproject" "secondeditor"
+            assertBool "Second claim result should fail." $ not secondClaimResult
+        , withTestPool "when a user claims control, they can claim it a second time immediately after" $ \pool -> do
+            metrics <- getDatabaseMetrics
+            firstClaimResult <- maybeClaimCollaborationControl metrics pool sixteenthOfJanuaryMorning "testowner" "testproject" "testeditor"
+            assertBool "First claim result should be successful." firstClaimResult
+            secondClaimResult <- maybeClaimCollaborationControl metrics pool sixteenthOfJanuaryMorning "testowner" "testproject" "testeditor"
+            assertBool "Second claim result should be successful." secondClaimResult
+        , withTestPool "when a user has claimed control, another can only claim it once the expiry has been exceeded" $ \pool -> do
```

Comment:
is owner here based on userID? so if I have the same project open in multiple tabs, that is multiple instances of "firstowner", right? so this "secondowner" is theoretial in the sense that the UI is not exposing this in any way, right? as in: a project in Utopia today cannot have two owners

Reply:
Yes, owner is user ID. I should probably change this to user. In this case the check for project owner is done at the level above this logic, so the call would never be made otherwise with a non-owner user ID.